### PR TITLE
Raw endpoint responds with 404 not found for cloning repo.

### LINKED ISSUE #19096: Reduce coupling of raw endpoint with repo-updater
### Problem description

If the `raw` endpoint for a particular repo receives a request and the repo is not cloned already, the [serveRaw](https://github.com/sourcegraph/sourcegraph/blob/main/cmd/frontend/internal/app/ui/raw.go#L62-L85) handler keeps checking indefinitely waiting for the repo to be cloned. This has the following down sides:

1. The client waits while the request is completed. 
1. If the `repo-updater` component is not available for any reason, the request will stay alive indefinitely until the max TTL of the web server is hit.
1. The service serving the `raw` endpoint is heavily coupled with the `repo-updater` and makes the service implicitly trust that the availability of `repo-updater`. Services should be fault tolerant and expect other services to fail and handle the failure scenarios appropriately.
1. This potentially opens up a DoS attack vector.

### Proposed solution

Instead of waiting for the repo to be cloned, we should return a 404 with a [useful error message](https://tools.ietf.org/html/draft-ietf-httpbis-p2-semantics-12#section-8.4) that indicates to the client that they may try again in some time and maybe the repo will be available then.

The scope of this issue should also include tests for the new behavior. 

### Dependencies with other areas of the product

- https://sourcegraph.com/github.com/sourcegraph/src-cli/-/blob/internal/batches/repo_fetcher.go#L291-338 - Owned by @sourcegraph/batchers

### PR DESCRIPTION
**Pull request by cbart**
Resolves #19096

Also updates a docstring on `newCommon` which referenced `RevSpec` symbol that does not seem searchable within the scope. Rather `newCommon` returns a `nil` `Repo` when cloning.

## Test plan

* A unit test expecting a 404 with mocked out repo updater through newCommon (which other tests in the directory seem to be using).

## Haiku

```
  Cloning status code
  Clear availability
  Do not need to wait
```

### CODE REVIEW

File: cmd/frontend/internal/app/ui/raw_test.go [hunk 1]
```diff
@@ -376,3 +378,45 @@ c.go`
...
 	})
 }
+
+func Test_serveRawRepoCloning(t *testing.T) {
+	mockNewCommon = func(w http.ResponseWriter, r *http.Request, title string, serveError serveErrorHandler) (*Common, error) {
+		return &Common{
+			Repo: nil,
+		}, nil
+	}
+	defer func() {
```
Comment (sashaostrikov) [inline cmd/frontend/internal/app/ui/raw_test.go:388]: nit: `t.Cleanup` can be used

File: cmd/frontend/internal/app/ui/raw_test.go [hunk 2]
```diff
@@ -376,3 +378,45 @@ c.go`
...
+	}()
+	select {
+	case <-done:
+		if err != nil {
+			t.Fatalf("Failed to invoke serveRaw: %v", err)
+		}
+	case <-time.After(2 * time.Second):
+		t.Fatalf("timed out waiting for repo raw")
+	}
+	if got, want := w.Code, http.StatusNotFound; got != want {
```
Comment (sashaostrikov) [inline cmd/frontend/internal/app/ui/raw_test.go:416]: you can import `"github.com/google/go-cmp/cmp"` and use it to check for difference between expected and actual results like this:
```go
if diff := cmp.Diff(http.StatusNotFound, http.StatusNotFound); diff != "" {
    t.Errorf("HTTP status mismatch (-want +got):\n%s", diff)
}
```

**Review by sashaostrikov — APPROVED** [review]
LGTM! :shipit:
Added a couple of suggestions for a test file.

**Review by mrnugget — APPROVED** [review]

File: cmd/frontend/internal/app/ui/raw_test.go (hunk 2 above)
Comment (indradhanush) [reply cmd/frontend/internal/app/ui/raw_test.go:416]: I think not using `cmp.Diff` in this case is cleaner. Using it makes sense for `structs` though.

Comment (indradhanush) [reply cmd/frontend/internal/app/ui/raw_test.go:416]: Why not directly do: `if w.Code != http.StatusNotFound`? 🤔

Comment (ryanslade) [reply cmd/frontend/internal/app/ui/raw_test.go:416]: I've actually been using the assert package a lot more recently. You can write something like

```go
assert.Equal(t, http.StatusNotFound, w.Code)
```

And it'll fail when not equal with a nice output. It's similar to `cmp.Diff` but easier to write.

File: cmd/frontend/internal/app/ui/raw.go [hunk 3]
```diff
@@ -62,29 +62,33 @@ import (
...
-				time.Sleep(5 * time.Second)
-				continue
-			}
-			break
+	return func(w http.ResponseWriter, r *http.Request) error {
+		const (
+			textPlain       = "text/plain"
+			applicationZip  = "application/zip"
+			applicationXTar = "application/x-tar"
+		)
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply cmd/frontend/internal/app/ui/raw.go:70]: Or maybe it's better to have it within the closure. No big deal.

File: cmd/frontend/internal/app/ui/raw_test.go (hunk 2 above)
Comment (indradhanush) [reply cmd/frontend/internal/app/ui/raw_test.go:416]: +1 for using assert! 

Note to self: I should be using these more in my tests.

File: cmd/frontend/internal/app/ui/raw_test.go (hunk 1 above)
Comment (cbart) [reply cmd/frontend/internal/app/ui/raw_test.go:388]: 👍

File: cmd/frontend/internal/app/ui/raw_test.go [hunk 4]
```diff
@@ -376,3 +378,45 @@ c.go`
...
+	gsClient := gitserver.NewMockClient()
+	gsClient.StatFunc.SetDefaultReturn(nil, fmt.Errorf("should not be invoked"))
+
+	req := httptest.NewRequest("GET", "/github.com/sourcegraph/sourcegraph/-/raw", nil)
+	w := httptest.NewRecorder()
+
+	db := database.NewMockDB()
+	// Former implementation would sleep awaiting repository to be available.
+	// Await request to be served with a timeout by racing done channel with time.After.
+	done := make(chan struct{})
```
[the remark this answers is not in the document]
Comment (cbart) [reply cmd/frontend/internal/app/ui/raw_test.go:402]: 👍

File: cmd/frontend/internal/app/ui/raw_test.go (hunk 2 above)
Comment (cbart) [reply cmd/frontend/internal/app/ui/raw_test.go:416]: 🤔 Definitely controversial suggestion. I'm happy to implement that and see how it works.

File: cmd/frontend/internal/app/ui/raw.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cbart) [reply cmd/frontend/internal/app/ui/raw.go:70]: Yep, I don't have a strong preference here either. It was kindof nice when it was defined literally next to the usage, but now it's a little more scattered. I think the only reason to keep it inside is scope, so let's leave it in.

File: cmd/frontend/internal/app/ui/raw_test.go [hunk 5]
```diff
@@ -376,3 +378,30 @@ c.go`
...
+func Test_serveRawRepoCloning(t *testing.T) {
+	mockNewCommon = func(w http.ResponseWriter, r *http.Request, title string, serveError serveErrorHandler) (*Common, error) {
+		return &Common{
+			Repo: nil,
+		}, nil
+	}
+	t.Cleanup(func() {
+		mockNewCommon = nil
+	})
+	// Fail git server calls, as they should not be invoked for a cloning repo.
```
Comment (indradhanush) [inline cmd/frontend/internal/app/ui/raw_test.go:391]: 😛

```suggestion
	// Fail gitserver calls, as they should not be invoked for a cloning repo.
```

**Review by indradhanush — APPROVED** [review]
🚢

**Pull request merged**